packit/upstream.py

```diff
@@ -248,22 +248,63 @@ def create_patches(
             self.package_config.patch_generation_ignore_paths + sync_files_to_ignore
         )
 
+        linearize_history = False
+        # let's check whether some merge commits have parents past <upstream>
+        commits = self.get_commits_to_upstream(
+            upstream, add_upstream_head_commit=True, no_merge_commits=False
+        )
+        for commit in commits[1:]:  # 0 = upstream, don't check that one
+            for parent_commit in commit.parents:
+                if parent_commit not in commits:
+                    logger.info(f"Commit {commit} has a parent past {upstream}.")
+                    logger.info(
+                        "Git may be unable to construct the list of patches for complex histories."
+                    )
+                    logger.info(
+                        "We are going to linearize the history on a dedicated branch."
+                    )
+                    linearize_history = True
+                    break
+            if linearize_history:
+                break
+        switch_back_to: Optional[str] = None
+        if linearize_history:
+            # if the git history is swarmed with merge commits,
+            # `git format-patch` tends to create patches which can't be applied
+            # in such a case, we should "linearize" the history
+            switch_back_to = run_command(
```

Comment:
Couldn't this whole mechanism be replaced with `git checkout -`?

Reply:
yes